File: solutions/c/isogram/1/isogram.c
#include "isogram.h"
#include <ctype.h>
#include <stdint.h>

bool is_isogram(const char* phrase) {
    if (!phrase) return false;
    uint32_t seen = 0;

    for (; *phrase; phrase++) {
        if (!isalpha(*phrase)) continue;
        int bit = 1 << (tolower(*phrase) - 'a');
        if (seen & bit) return false;
        seen |= bit;
    }
    return true;
}

File: solutions/haskell/isogram/4/src/Isogram.hs
module Isogram (isIsogram) where

import Data.List (nubBy)
import Data.Char (isAlpha, toLower)

isIsogram :: String -> Bool
isIsogram phrase = nubBy equalLetters phrase == phrase
  where
    equalLetters c1 c2 = isAlpha c1 &&
      toLower c1 == toLower c2
